xdg-launch.1
.TH XDG-LAUNCH 1 "April 2025" "1.6" "XDG Launcher"

.SH NAME
xdg-launch \- Enforces XDG Base Directory compliance for applications.

.SH SYNOPSIS
.B xdg-launch
[\fB\-\-cd\fR \fIdirectory\fR]
[\fB\-\-fuzz\fR]
[\fB\-\-quiet\fR]
.IR application
[\fB\-\- \fIoptions\fR...]

.SH DESCRIPTION
.B xdg-launch
is a shell script designed to enforce the XDG Base Directory Specification for
applications that do not respect the user's XDG settings. By temporarily
redefining the \fIHOME\fR environment variable and linking necessary files,
it fools applications into storing their configuration, cache, and data files
in the appropriate XDG-defined locations.

.SH OPTIONS

.TP
.B \-d, \--dir, \--cd \fIdirectory\fR
Specify a working directory for the application.
The script will change to this directory before launching the application.

.TP
.B \-f, \--fuzz
Fuzz the \fI/etc/passwd\fR file for applications that ignore environment variables.
This enables namespace isolation and maps the fake root user's home directory to \fIXDG_DATA_HOME\fR.
This ensures compatibility for applications that target on \fI/etc/passwd\fR for determining the home directory.

.TP
.BR \-\-help ", " \-h
Display usage information and exit.

.TP
.BR \-\-quiet ", " \-q
Quiet mode to supress the launcher messages. The launched application's output is not affected.

.TP
.I application
The application to be launched. This is required.

.TP
.I options
Any additional options or arguments passed to the application must follow the 
.B --
separator.

.SH HOW IT WORKS

.TP
Environment Setup:
The script checks for user-defined \fBXDG_CONFIG_HOME\fR, \fBXDG_CACHE_HOME\fR,
\fBXDG_DATA_HOME\fR and \fBXDG_STATE_HOME\fR settings.
If they are not defined, default values are used.
Missing directories are created as needed.

.TP
Resource Useage Registratoin:
The quantity of applications using links within the specified \fBXDG_DATA_HOME\fR is
saved within the \fI$XDG_RUNTIME_DIR/xdg/launcher.data\fR file.
A a mutex lock on the \fI$XDG_RUNTIME_DIR/xdg/launcher.lock\fR file is used to ensure
only one instance of the launcher updates this data at any time.

.TP
Temporary Links:
Temporary links are created within \fBXDG_DATA_HOME\fR for the user's
cache, config, and Xauthority (if necessary).

.TP
/etc/passwd Fuzzing (if enabled):
A copy of the \fI/etc/passwd\fR file is created as \fI$XDG_RUNTIME_DIR/xdg/passwd.pid\fR,
where \fBpid\fR is the process id of the launcher.
Within this file, the root user's home directory is set to the real user's \fBXDG_DATA_HOME\fR.

.TP
Application Launch:
The script changes the working directory (if specified), temporarily redefines the
\fBHOME\fR environment variable to point to \fBXDG_DATA_HOME\fR, and
launches the specified application with its arguments.
If enabled, this occurs within a new private namespace for fuzzing the \fI/etc/passwd\fR file.

.TP
Cleanup After the Application Exits:
The number of registerd applications is decremented within the \fI$XDG_RUNTIME_DIR/xdg/launcher.data\fR file.
Temporary links are removed unless they are still in use by another application.
Fuzzed passwd files are removed from the \fI$XDG_RUNTIME_DIR/xdg\fR directory.

.SH ENVIRONMENT VARIABLES

.TP
.B XDG_CONFIG_HOME
Defines where user-specific configuration files should be stored. Defaults to 
.I ${HOME}/.config.

.TP
.B XDG_CACHE_HOME
Defines where user-specific cache files should be stored. Defaults to 
.I ${HOME}/.cache.

.TP
.B XDG_DATA_HOME
Defines where user-specific data files should be stored. Defaults to 
.I ${HOME}/.local/share.

.TP
.B XAUTHORITY
Specifies the path to the \fI.Xauthority\fR file used by X11 for authentication.

.SH NOTES
.PP
If you have an existing \fBXAUTHORITY\fR setting, the script will \fBnot\fR
create a temporary link for \fI.Xauthority\fR. It is expected that your setting
works without requiring the link.
Please let the author know if this is, in fact, true or not.

.SH EXIT STATUS
.PP
The launcher returns the exit code of the target application.

.SH FILES
.PP
The following directories and files are used:
.TP
.I $HOME/.config:
Default location for configuration files.
.TP
.I $HOME/.cache:
Default location for cache files.
.TP
.I $HOME/.local/share:
Default location for data files.
.TP
.I $HOME/.local/state:
Default location for state data files.
.TP
.I $HOME/.Xauthority:
Used for X11 authentication.
.TP
.I $XDG_RUNTIME_DIR/xdg/launcher.data
Resource registry to track number of applications currently using the launcher.
.TP
.I $XDG_RUNTIME_DIR/xdg/launcher.lock
A lock file to ensure mutual exclusion for acessing the launcher data registry.
.TP
.I $XDG_RUNTIME_DIR/xdg/passwd.pid
Any fuzzed \fIpasswd\fR files are placed within the \fI$XDG_RUNTIME_DIR/xdg\fR directory.
They are differentiated by the process id, \fIpid\fR, of the launcher which uses it.

.SH AUTHOR

Hildigerr Vergaray  

.SH SEE ALSO

.XDG Base Directory Specification:
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html  



xdg-launch.sh
#!/bin/sh
#
# Launcher Enforcing XDG Base Directory Compliance
#
#  Some applications refuse to comply with user's XDG Base Directory settings.
#  So, we redefine HOME to enforce compliance.
#  However, they often need to be able to find the Xauthority file.
#  So, we temporarily link it.
#  Some applications half-ass comply using cache and/or config for some things.
#  So, we temporarily link them too, enforcing XDG_{CACHE,CONFIG}_HOME setting.
#
# Copyright (C) 2025 Hildigerr Vergaray [MIT License]

usage() { echo "Usage: $0 [-cd <directory>] [--fuzz] [--quiet] application> [-- options]"; exit; }

increment_lock() {
  touch "${STATE_FILE}"
  local update="$(awk -F'::' -v OFS='::' -v path="${XDG_DATA_HOME}" '
      BEGIN { count = 0 }
      $2 == path { count = ++$1 }
      { print }
      END { if (!count) { print "1::" path } }
    ' "${STATE_FILE}")" || {
      echo "Error: Failed to register resource lock" >&2
      exit 1
    }
  echo "$update" > "${STATE_FILE}"
}

decrement_lock() {
  local count=$(awk -F'::' -v path="${XDG_DATA_HOME}" '
    $2 == path { print $1; exit }
  ' "${STATE_FILE}" ) || count=0

  if [ "$count" -lt 1 ]; then
    echo "Error: Resource lock corrupted [${count}]" >&2
    echo "       Unlinking will not occur in case links are still needed." >&2
    exit 1
  fi

  sed -i "s|$count::${XDG_DATA_HOME}$|$((count - 1))::${XDG_DATA_HOME}|" "${STATE_FILE}"
  if [ ! $? ]; then
    echo "Error: Failed to deregister resource lock [$count]" >&2
    echo "       Unlinking will not occur in case links are still needed." >&2
    exit 1
  fi
  return $((count - 1))
}

# Save stdout and stderr for quiet mode
exec 3>&1 4>&2

while [ "$#" -gt 0 ]; do
  case $1 in
    --dir|--cd|-d)
      WORKING_DIRECTORY="$2"
      if [ -d "$WORKING_DIRECTORY" ]; then
        pushd "$WORKING_DIRECTORY"
      else
        echo "Error: No such directory: \"$WORKING_DIRECTORY\""
        exit 1
      fi
      shift;;
    --fuzz|-f) FUZZ="${XDG_RUNTIME_DIR:-/tmp/$USER}/xdg/passwd.$$";;
    --help|-h) useage;;
    --quiet|-q) QUIET="$1";;
    --) shift; break;;
    *) if [ -z "$APPLICATION" ]; then APPLICATION="$1"; else usage; fi;;
  esac
  shift
done
if [ -z "$APPLICATION" ]; then usage; fi;

if [ -n "$QUIET" ]; then exec >/dev/null 2>&1; fi

# Determine Where User Config Files Should Go
XDG_CONFIG_HOME="${XDG_CONFIG_HOME:-${HOME}/.config}"
echo "XDG_CONFIG_HOME is \"${XDG_CONFIG_HOME}\""

# Create User Config Home (Should be Unnecessary)
if [ ! -d "${XDG_CONFIG_HOME}" ]; then
  echo "Creating \"${XDG_CONFIG_HOME}\""
  mkdir -p "${XDG_CONFIG_HOME}"
fi

# Determine Where User Cache Should Go
XDG_CACHE_HOME="${XDG_CACHE_HOME:-${HOME}/.cache}"
echo "XDG_CACHE_HOME is \"${XDG_CACHE_HOME}\""

# Create User Cache Home (Should be Unnecessary)
if [ ! -d "${XDG_CACHE_HOME}" ]; then
  echo "Creating \"${XDG_CACHE_HOME}\""
  mkdir -p "${XDG_CACHE_HOME}"
fi

# Determine Where User Data Files Should Go
export XDG_DATA_HOME=${XDG_DATA_HOME:-${HOME}/.local/share}
echo "XDG_DATA_HOME is \"${XDG_DATA_HOME}\""

# Create User Data Home (Should be Unnecessary)
if [ ! -d "${XDG_DATA_HOME}" ]; then
  echo "Creating \"${XDG_DATA_HOME}\""
  mkdir -p "${XDG_DATA_HOME}"
fi

# Determine Where User State Date Should Go
XDG_STATE_HOME=${XDG_STATE_HOME:-${HOME}/.local/state}
echo "XDG_STATE_HOME is \"${XDG_STATE_HOME}\""

STATE_HOME="${XDG_RUNTIME_DIR:-$XDG_STATE_HOME}/xdg"

# Create User State Home for Launcher
if [ ! -d "${STATE_HOME}" ]; then
  echo "Creating \"${STATE_HOME}\""
  mkdir -p "${STATE_HOME}"
fi

# Register State as in-use
LOCK_FILE="${STATE_HOME}/launcher.lock"
export STATE_FILE="${STATE_HOME}/launcher.data"
echo "Acquiring resource lock..."
flock -x "${LOCK_FILE}" -c "$(declare -f increment_lock); increment_lock"

# Link Xauthority file (Only if it Known to be Necessary)
if [ -n "$XAUTHORITY" ] && [ -f "$XAUTHORITY" ]; then
  # User has XAUTHORITY setting, lets see if it works withough a cludge.
  echo "XXX: NOT temporarily linking your XAUTHORITY file."
  echo "     Please, let the maintainer know if this works."
elif [ -f "${HOME}/.Xauthority" ] && [ ! -f "${XDG_DATA_HOME}/.Xauthority" ]; then
  echo "Linking ~/.Xauthority within \"${XDG_DATA_HOME}\""
  ln --symbolic -T "${HOME}/.Xauthority" "${XDG_DATA_HOME}/.Xauthority"
fi

# Link User's Cache Home
if [ ! -e "${XDG_DATA_HOME}/.cache" ]; then
  echo "Linking User's Cache Home..."
  ln --symbolic -T "${XDG_CACHE_HOME}" "${XDG_DATA_HOME}/.cache"
fi

# Link User's Config Home
if [ ! -e "${XDG_DATA_HOME}/.config" ]; then
  echo "Linking User's Config Home..."
  ln --symbolic -T "${XDG_CONFIG_HOME}" "${XDG_DATA_HOME}/.config"
fi

if [ -n "$QUIET" ]; then exec >&3 2>&4; fi

if [ -z "${FUZZ}" ]; then
  env HOME="${XDG_DATA_HOME}" "$APPLICATION" "$@"
else
  mkdir -p "$(dirname "${FUZZ}")"
  sed "s|:/root:|:${XDG_DATA_HOME}:|" /etc/passwd > "${FUZZ}"
  unshare --user --map-root-user --mount --propagation private sh -c "
    mount --bind \"${FUZZ}\" /etc/passwd
    exec env HOME=\"${XDG_DATA_HOME}\" \"${APPLICATION}\" \"\$@\"
  " _ "$@"
fi
RVAL=$?

if [ -n "${FUZZ}" ]; then unlink "${FUZZ}"; fi

if [ -n "$QUIET" ]; then exec >/dev/null 2>&1; fi

echo "Releasing resource lock..."
flock -x "${LOCK_FILE}" -c "$(declare -f decrement_lock); decrement_lock"
if [ $? -gt 0 ]; then
  echo "Resources still in use. Not unlinking."
  exit $RVAL
fi

# Unlink the Xauthority file (Only if it is symbolic)
if [ -L "${XDG_DATA_HOME}/.Xauthority" ]; then
  echo "Unlinking temporary Xauthority file"
  unlink "${XDG_DATA_HOME}/.Xauthority"
fi

# Unlink User's Cache Home
if [ -L "${XDG_DATA_HOME}/.cache" ]; then
  echo "Unlinking temporary cache home"
  unlink "${XDG_DATA_HOME}/.cache"
fi

# Unlink User's Config Home
if [ -L "${XDG_DATA_HOME}/.config" ]; then
  echo "Unlinking temporary config home"
  unlink "${XDG_DATA_HOME}/.config"
fi

exit $RVAL
